usr/lib/common/pin_get.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * COPYRIGHT (c) International Business Machines Corp. 2022
+ *
+ * This program is provided under the terms of the Common Public License,
+ * version 1.0 (CPL-1.0). Any use, reproduction or distribution for this
+ * software constitutes recipient's acceptance of CPL-1.0 terms which can be
+ * found in the file LICENSE file or at
+ * https://opensource.org/licenses/cpl1.0.php
+ */
+
+#define _POSIX_C_SOURCE 200809L
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <termios.h>
+#include <unistd.h>
+#include <sys/mman.h>
+
+#include <pin_get.h>
+
+struct pinbuf pinbuf_empty = {
+    .p = NULL,
+    .len = 0,
+};
+
+static void echo(bool on)
+{
+    struct termios term;
+
+    tcgetattr(fileno(stdin), &term);
+
+    if (on)
+        term.c_lflag |= ECHO;
+    else
+        term.c_lflag &= ~ECHO;
+
+    tcsetattr(fileno(stdin), TCSAFLUSH, &term);
+}
+
+const char *pin_prompt(struct pinbuf *buf, const char *msg)
+{
+    ssize_t n;
+
+    if (!buf)
+        return NULL;
+
+    printf("%s", msg);
+    fflush(stdout);
+    echo(false);
+
+    n = getline(&buf->p, &buf->len, stdin);
+
+    echo(true);
+    printf("\n");
+    fflush(stdout);
+
+    /* delayed getline() error handling */
+    if (n == -1)
+        return NULL;
+
+    /* strip CR/LF (if any) */
+    (buf->p)[strcspn(buf->p, "\r\n")] = '\0';
+
+    buf->len = strlen(buf->p);
```

Comment:
If the pin would contain a \n somewhere in the middle (not sure if getline would allow that at all), then you would get the wrong length here.
So I again wonder why you need the pinbuf struct at all, if the length is set via strlen anyway..

Reply:
Yes and no. The pin should not contain \0, but the length calculation is anyway wrong. Len holds the allocated length, not the length of the pin. I'll fix that.